### PR TITLE
NIPagingScrollView additions - much simpler inset implementation

### PR DESCRIPTION
**Pull request by georgeharker**
NIPagingScrollView.pageInset if non zero will inset the page from full screen, thus making neighboring pages visible.  This is useful for showing that there is scrollable content and presents a 'film-strip' like appearance.

### CODE REVIEW

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 1]
```diff
@@ -393,6 +421,11 @@ - (void)setFrame:(CGRect)frame {
   [self layoutVisiblePages];
 }
 
+- (void)layoutSubviews {
+  _scrollView.contentSize = [self contentSizeForPagingScrollView];
+  [self layoutVisiblePages];
```
Comment (jverkoey) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Why do we need to do this now?

You should also _always_ call `[super layoutSubviews]`; Not doing so can break things like autolayout constraints (and so also autoresizing masks).

Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: I thought we did, because the combo of setting the content size and invoking layoutVisiblePages is definitely required.  However, I can move that to setPageInset and that's a lot simpler.

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 2]
```diff
@@ -195,7 +198,7 @@ - (CGFloat)scrolledPageOffset {
 - (BOOL)isDisplayingPageForIndex:(NSInteger)pageIndex {
   BOOL foundPage = NO;
 
-  // There will never be more than 3 visible pages in this array, so this lookup is
+  // There will never be more than 3 (5 with insets) visible pages in this array, so this lookup is
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.m]: The number of visible pages is not constant with a non-zero page inset; it is a function of the page inset, page size, and viewport.

File: src/pagingscrollview/src/NIPagingScrollView.h [hunk 3]
```diff
@@ -68,7 +68,11 @@ typedef enum {
 
 #pragma mark Configuring Presentation
 
+// Controls the border between images.
 @property (nonatomic) CGFloat pageMargin;
+// Used to make the view smaller than full screen, thus showing neighboring images
+// in the view.
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.h]: You probably want to make it clear that the page inset is only horizontal.

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 4]
```diff
@@ -64,6 +66,7 @@ - (void)commonInit {
   _scrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
   _scrollView.pagingEnabled = YES;
   _scrollView.scrollsToTop = NO;
+  _scrollView.clipsToBounds = NO;
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Is this change actually needed?

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 5]
```diff
@@ -694,6 +722,20 @@ - (void)setPageMargin:(CGFloat)pageMargin {
   [self setNeedsLayout];
 }
 
+- (void)setPageInset:(CGFloat)pageInset {
+  _pageInset = pageInset;
+  _scrollView.frame = [self frameForPagingScrollView];
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.m]: It seems odd for the page inset setter to modify the scroll view's frame. I would expect the setter to invalidate the layout like setPageMargin: above. On a related note I get the sense that this class is missing an implementation of layoutSubviews?

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 6]
```diff
@@ -227,15 +230,35 @@ - (NSInteger)currentVisiblePageIndex {
   return 0;
 }
 
+- (NSInteger)pageIndexForOffset:(CGFloat)offset {
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Is this method actually used?

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 4 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Yes, we rely on showing the bleed underneath the bounds to show the neighboring images.

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 5 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Funny, I had one originally.  I got some comments about it so removed it.  But it's easy to add back ;)

File: src/pagingscrollview/src/NIPagingScrollView.h (hunk 3 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.h]: Sure.  Have done.

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 7]
```diff
@@ -350,15 +373,20 @@ - (void)updateVisiblePagesShouldNotifyDelegate:(BOOL)shouldNotifyDelegate {
 
     [self didChangeCenterPageIndexFrom:oldCenterPageIndex to:_centerPageIndex];
 
-    // Prioritize displaying the currently visible page.
-    if (![self isDisplayingPageForIndex:_centerPageIndex]) {
-      [self displayPageAtIndex:_centerPageIndex];
+    if (_pageInset != 0) {
+      // When multiple images are visible load them all.
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.m]: This comment seems inaccurate and doesn't reflect the actual condition it's based upon. Maybe it should read "Load all visible insetted pages immediately."?

File: src/pagingscrollview/src/NIPagingScrollView.h (hunk 3 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.h]: actually, it's not, it does vertical offset for vertical mode.

File: src/pagingscrollview/src/NIPagingScrollView.h [hunk 8]
```diff
@@ -68,7 +68,11 @@ typedef enum {
 
 #pragma mark Configuring Presentation
 
+// Controls the border between images.
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.h]: s/images/pages/

Also may want to mention that the page margin is only applied in the paging direction.

File: src/pagingscrollview/src/NIPagingScrollView.h [hunk 9]
```diff
@@ -68,7 +68,11 @@ typedef enum {
...
 @property (nonatomic) CGFloat pageMargin;
+// Used to make the view smaller than full screen, thus showing neighboring images
```
Comment (stephanemoore) [inline src/pagingscrollview/src/NIPagingScrollView.h]: s/images/pages/

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 2 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Agreed.  Reworking that.

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 7 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: just making those changes.

File: src/pagingscrollview/src/NIPagingScrollView.h (hunk 8 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.h]: done.

File: src/pagingscrollview/src/NIPagingScrollView.h (hunk 9 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.h]: Fixed that.

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 4 above)
Comment (jverkoey) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Perhaps we can enable this only if insets are enabled?

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 2 above)
Comment (jverkoey) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Comment's still out of date.

**Comment by jverkoey** [comment]
Testing on the sim reveals that, due to futzing with the scroll view's frame, you can't drag it from the edges of the screen. You should be able to resolve this by simply forwarding touch events that land on the paging scroll view to the UIScrollView itself.

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 4 above)
Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: I just checked and I believe the default is actually clipsToBounds=NO, but I was trying to be explicit about the need for it.

File: src/pagingscrollview/src/NIPagingScrollView.m [hunk 10]
```diff
@@ -393,6 +420,13 @@ - (void)setFrame:(CGRect)frame {
   [self layoutVisiblePages];
 }
 
+- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
```
Comment (jverkoey) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Add a comment explaining why we have to do this - otherwise future maintainer will have to figure it out themselves :(

Comment (georgeharker) [inline src/pagingscrollview/src/NIPagingScrollView.m]: Done.

**Comment by jverkoey** [comment]
Thanks! This is awesome :) We'll keep an eye on layoutSubviews to see if it breaks any existing implementations.

File: src/pagingscrollview/src/NIPagingScrollView.m (hunk 4 above)
Comment (jverkoey) [inline src/pagingscrollview/src/NIPagingScrollView.m]: The default is actually YES.
